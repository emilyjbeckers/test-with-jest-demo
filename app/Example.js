import React from 'react';
import API from './api';
import {
  Button,
  StyleSheet,
  Text,
  View,
} from 'react-native';


export default class Example extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      annoyed: false,
    };
    this.data = 'Not yet...';
  }

  componentDidMount() {
    this._fetchData();
  }

  async _fetchData() {
    this.data = await API.fetch();
  }

  _onPress = () => {
    this.setState({annoyed: true});
  }

  render() {
    return (
      <View style={[styles.main]}>
        <Text>Is it ready? Press the button to find out.</Text>
        <Text>{this.data}</Text>
        <Button
          onPress={this._onPress}
          title={this.state.annoyed ? 'That\'s enough.' : 'Press me!'}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  main: {
    height: 300,
    justifyContent: 'space-around',
    backgroundColor: '#ccff99',
  },
});
