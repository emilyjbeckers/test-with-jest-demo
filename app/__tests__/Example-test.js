import 'react-native';
import React from 'react';

import API from '../api';
import Example from '../Example';
import renderer from 'react-test-renderer';

jest.useFakeTimers();

jest.mock('../api', () => {
  return {
    fetch: jest.fn().mockResolvedValue('some very cool data'),
  };
});

describe('Example', () => {
  it('renders correctly when not annoyed', () => {
    const example = renderer.create(<Example />);
    expect(example.getInstance().state.annoyed).toBeFalsy();
    expect(example.toJSON()).toMatchSnapshot();
  });

  it('renders correctly when annoyed', (done) => {
    expect.assertions(1);

    const example = renderer.create(<Example />);
    example.getInstance().setState({annoyed: true}, () => {
      expect(example.toJSON()).toMatchSnapshot();
      done();
    });
  });
});

describe('componentDidMount', () => {
  it('calls _fetchData', () => {
    const example = renderer.create(<Example />).getInstance();

    const spy = jest.spyOn(example, '_fetchData');
    expect(spy).not.toHaveBeenCalled();

    example.componentDidMount();

    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });
});

describe('_onPress', () => {
  it('updates state.annoyed to true', () => {
    const example = renderer.create(<Example />).getInstance();

    expect(example.state).toHaveProperty('annoyed', false);

    const spy = jest.spyOn(example, 'setState');
    expect(spy).not.toHaveBeenCalled();

    example._onPress();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith({annoyed: true});

    spy.mockRestore();
  });
});


describe('_fetchData', () => {
  it('calls API.fetch and updates the data when fetch resolves', (done) => {
    expect.assertions(4);
    const example = renderer.create(<Example />).getInstance();
    API.fetch.mockClear();
    expect(example.data).toEqual('Not yet...');
    expect(API.fetch).not.toHaveBeenCalled();
    API.fetch.mockResolvedValueOnce('It works!');

    example._fetchData().then(() => {
      expect(API.fetch).toHaveBeenCalled();
      expect(example.data).toEqual('It works!');
    }).finally(() => {
      API.fetch.mockClear();
      done();
    });
  });
});
