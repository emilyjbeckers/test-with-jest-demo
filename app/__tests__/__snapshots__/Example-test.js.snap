// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example renders correctly when annoyed 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#ccff99",
        "height": 300,
        "justifyContent": "space-around",
      },
    ]
  }
>
  <Text>
    Is it ready? Press the button to find out.
  </Text>
  <Text>
    Not yet...
  </Text>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        That's enough.
      </Text>
    </View>
  </View>
</View>
`;

exports[`Example renders correctly when not annoyed 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#ccff99",
        "height": 300,
        "justifyContent": "space-around",
      },
    ]
  }
>
  <Text>
    Is it ready? Press the button to find out.
  </Text>
  <Text>
    Not yet...
  </Text>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Press me!
      </Text>
    </View>
  </View>
</View>
`;
